pkg/dns/target.go

```diff
@@ -1,11 +1,186 @@
+type ClusterGateway struct {
+	ClusterName       string
+	GatewayAddresses  []gatewayv1beta1.GatewayAddress
+	ClusterAttributes ClusterAttributes
+}
+
+type ClusterAttributes struct {
+	Geo *GeoCode
+}
+
+type GeoCode string
+
+// IsContinentCode returns true if it's a valid continent code
+func (gc GeoCode) IsContinentCode() bool {
+	return slice.ContainsString(getContinentCodes(), string(gc))
+}
+
+// IsCountryCode returns true if it's a valid ISO_3166 country code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)
+func (gc GeoCode) IsCountryCode() bool {
+	//ToDo implement me !!!
+	return true
+}
+
+func (gc GeoCode) IsDefaultCode() bool {
+	return gc == DefaultGeo
+}
+
+func (gc GeoCode) IsValid() bool {
+	return gc.IsContinentCode() || gc.IsCountryCode()
+}
+
+// AF: Africa; AN: Antarctica; AS: Asia; EU: Europe; OC: Oceania; NA: North America; SA: South America
+func getContinentCodes() []string {
+	return []string{"AF", "AN", "AS", "EU", "OC", "NA", "SA"}
+}
+
+func NewClusterGateway(cluster metav1.Object, gatewayAddresses []gatewayv1beta1.GatewayAddress) *ClusterGateway {
+	cgw := &ClusterGateway{
+		ClusterName:      cluster.GetName(),
+		GatewayAddresses: gatewayAddresses,
+	}
+	cgw.setClusterAttributesFromObject(cluster)
+	return cgw
+}
+
+func (t *ClusterGateway) setClusterAttributesFromObject(mc metav1.Object) {
+	t.ClusterAttributes = ClusterAttributes{}
+	for k, v := range mc.GetAnnotations() {
+		if k == LBAttributeAnnotationGeoCode {
+			geoCode := GeoCode(v)
+			if geoCode.IsValid() {
+				t.ClusterAttributes.Geo = &geoCode
+			}
+		}
+	}
+}
+
+// ClusterGatewayTarget represents a cluster Gateway with geo and weighting info calculated
+type ClusterGatewayTarget struct {
+	*ClusterGateway
+	Geo    *GeoCode
+	Weight *int
+}
+
+func NewClusterGatewayTarget(cg ClusterGateway) ClusterGatewayTarget {
+	target := ClusterGatewayTarget{
+		ClusterGateway: &cg,
+	}
+	target.setGeo(DefaultGeo)
+	target.setWeight(DefaultWeight)
+	return target
+}
+
+func (t *ClusterGatewayTarget) GetGeo() GeoCode {
+	return *t.Geo
+}
+
+func (t *ClusterGatewayTarget) GetWeight() int {
+	return *t.Weight
+}
+
+func (t *ClusterGatewayTarget) GetName() string {
+	return t.ClusterName
+}
+
+func (t *ClusterGatewayTarget) GetShortCode() string {
+	return toBase36hash(t.GetName())
+}
+
+func (t *ClusterGatewayTarget) setGeo(defaultGeo GeoCode) {
+	if defaultGeo == DefaultGeo || t.ClusterAttributes.Geo == nil {
+		t.Geo = &defaultGeo
+		return
+	}
+	t.Geo = t.ClusterAttributes.Geo
+}
+
+func (t *ClusterGatewayTarget) setWeight(defaultWeight int) {
+	t.Weight = &defaultWeight
+}
+
+func toBase36hash(s string) string {
+	hash := sha256.Sum224([]byte(s))
+	// convert the hash to base36 (alphanumeric) to decrease collision probabilities
+	base36hash := strings.ToLower(base36.EncodeBytes(hash[:]))
+	// use 6 chars of the base36hash, should be enough to avoid collisions and keep the code short enough
+	return base36hash[:6]
 }
```

Comment:
KCP were you using this to generate workspace and logical cluster names. Assuming they knew better than me what they were doing i think this should be sufficient, it may even be possible to reduce the length further,  4 maybe?